docbrown/db/src/algorithms/reciprocity.rs

```diff
@@ -0,0 +1,102 @@
+use itertools::Itertools;
+use crate::graph_window::{WindowedGraph, WindowedVertex};
+
+
+fn get_unique_out_neighbours(v: &WindowedVertex) -> Vec<u64> {
+    let mut all_out_neighbours: Vec<u64> = v.out_neighbours_ids().unique().collect();
+    // remove self-loop
+    if let Some(pos) = all_out_neighbours.iter().position(|x| *x == v.g_id) {
+        all_out_neighbours.remove(pos);
+    }
+    all_out_neighbours
+}
+
+fn get_unique_in_neighbours(v: &WindowedVertex) -> Vec<u64> {
+    let mut in_neighbours: Vec<u64>  = v.in_neighbours_ids().unique().collect();
+    // remove self-loop
+    if let Some(pos) = in_neighbours.iter().position(|x| *x == v.g_id) {
+        in_neighbours.remove(pos);
+    }
+    in_neighbours
+}
+
+fn get_reciprocal_edge_count(v: &WindowedVertex) -> usize {
+    let out_neighbours: Vec<u64> = get_unique_out_neighbours(&v);
+    let in_neighbours: Vec<u64> = get_unique_in_neighbours(&v);
+    out_neighbours.iter().filter(|x| in_neighbours.contains(x)).count()
+}
+
+pub fn global_reciprocity(graph: &WindowedGraph) -> f64 {
+    let num_unique_edges = graph
+        .vertices()
+        .map(|v|
+            get_unique_out_neighbours(&v).len()
+        ).sum::<usize>();
+
+    let reciprocal_edges = graph
+        .vertices()
+        .map(|v|
+            get_reciprocal_edge_count(&v)
+        ).sum::<usize>();
+    (reciprocal_edges as f64 / num_unique_edges as f64)
+}
+
+// Returns the reciprocity of every vertex in the grph as a tuple of
+// vector id and the reciprocity
+pub fn all_local_reciprocity(graph: &WindowedGraph) -> Vec<(u64, f64)> {
+    let mut res = graph
+        .vertices()
+        .map(|v|
+                 (v.g_id, local_reciprocity(&graph, v.g_id)))
+        .collect::<Vec<(u64, f64)>>();
+    res.sort_by(|a, b| a.0.cmp(&b.0));
```

Comment:
Probably ok to leave it unsorted, if a user wanted a top K t etc. we can implement in a different function.

Reply:
no it needs to be sorted, because then it wont match with my expected value